### PR TITLE
[meshcop] support non-standard TXT entries at runtime

### PR DESCRIPTION
**Pull request by wgtdkp**
See also https://android-review.git.corp.google.com/c/platform/external/ot-br-posix/+/3111060
for how this change is used and tested.

### CODE REVIEW

File: src/border_agent/border_agent.hpp [hunk 1]
```diff
@@ -161,6 +163,10 @@ class BorderAgent : private NonCopyable
     std::map<std::string, std::vector<uint8_t>> mMeshCopTxtUpdate;
```
Comment (superwhd) [inline src/border_agent/border_agent.hpp:163]: Instead of introducing `mNonStandardTxtEntries`, I'd suggest to reuse `mMeshCopTxtUpdate`

Comment (wgtdkp) [reply src/border_agent/border_agent.hpp:162]: done!

File: src/border_agent/border_agent.cpp [hunk 2]
```diff
@@ -149,20 +149,29 @@ BorderAgent::BorderAgent(otbr::Ncp::RcpHost &aHost, Mdns::Publisher &aPublisher)
...
+    {
+        VerifyOrExit(!txtEntry.mKey.empty() && txtEntry.mKey.front() == 'v', error = OTBR_ERROR_INVALID_ARGS);
+    }
 
     mProductName = aProductName;
     mVendorName  = aVendorName;
     mVendorOui   = aVendorOui;
+    for (const auto &txtEntry : aNonStandardTxtEntries)
+    {
+        mMeshCopTxtUpdate[txtEntry.mKey] = txtEntry.mValue;
```
Comment (superwhd) [inline src/border_agent/border_agent.cpp:173]: Shall we clear the `mMeshCopTxtUpdate` before adding the items?

**Review by superwhd — APPROVED** [review]
LGTM overall 👍 

In the future we may want to unify the two ways (D-BUS, Android overlay) for customizing meshcop TXT for better readability.

**Review by bukepo — COMMENTED** [review]
can we use  the existing HandleUpdateVendorMeshCoPTxtEntries?

**Comment by wgtdkp** [comment]
> can we use the existing HandleUpdateVendorMeshCoPTxtEntries?

That one doesn't support overriding the vendor or product name

**Comment by bukepo** [comment]
> > can we use the existing HandleUpdateVendorMeshCoPTxtEntries?
> 
> 
> 
> That one doesn't support overriding the vendor or product name

I think we should allow overriding these fields by dbus api as well.

**Comment by wgtdkp** [comment]
> > > can we use the existing HandleUpdateVendorMeshCoPTxtEntries?
> > 
> > 
> > That one doesn't support overriding the vendor or product name
> 
> I think we should allow overriding these fields by dbus api as well.

I am not sure that's a good idea. The `SetMeshCopServiceValues` method for Android is documented clearly that this must be called before BorderAgent is enabled, so that it won't cause a vendor or model name change which will cause unexpected behavior - mobile devices are depending on the names to take specific actions. Generally, it doesn't sound reasonable for a device to be Google Border Router at T1 and then changes to be Apple Border Router at T2.

It would require the dbus client side change if you want to ensure that advertised vendor name never changes for dbus-controlled device. But this is out of scope I think.

**Review by bukepo — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Pull request merged**